### PR TITLE
WIP : try to define a consistent SearchQuery.Builder that make sense

### PR DESCRIPTION
**Pull request by mbaechler**
It doesn't even compile but I tried to define a decent SearchQuery.Builder

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxQuery.java [hunk 1]
```diff
@@ -61,99 +42,186 @@ public static Builder builder(MailboxSession session) {
...
 
     /**
-     * Gets the base reference for the search.
+     * Gets the searchBase reference for the search.
      * 
-     * @return the base
+     * @return the searchBase
      */
     public final MailboxPath getBase() {
         return base;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxQuery.java]: We still have this mailboxPath?

File: mailbox/api/src/test/java/org/apache/james/mailbox/AbstractSubscriptionManagerTest.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
 import org.junit.Test;
 
 /**
- * Abstract base class to test {@link SubscriptionManager} implementations
+ * Abstract searchBase class to test {@link SubscriptionManager} implementations
```
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/AbstractSubscriptionManagerTest.java]: Oups

File: mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/AbstractJCRScalingMapper.java [hunk 3]
```diff
@@ -28,7 +28,7 @@
 import org.apache.james.mailbox.store.transaction.TransactionalMapper;
 
 /**
- * Abstract base class for Mapper's which support scaling. This supports Level 1 Implementations of JCR. So no real transaction management is used. 
+ * Abstract searchBase class for Mapper's which support scaling. This supports Level 1 Implementations of JCR. So no real transaction management is used.
```
Comment (chibenwa) [inline mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/AbstractJCRScalingMapper.java]: oups

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/JPAMessageManager.java [hunk 4]
```diff
@@ -47,7 +47,7 @@
 import org.apache.james.mailbox.store.search.MessageSearchIndex;
 
 /**
- * Abstract base class which should be used from JPA implementations.
+ * Abstract searchBase class which should be used from JPA implementations.
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/JPAMessageManager.java]: oups

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 5]
```diff
@@ -67,7 +67,7 @@
 import com.google.common.base.Objects;
 
 /**
- * Abstract base class for JPA based implementations of
+ * Abstract searchBase class for JPA based implementations of
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: oups

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java [hunk 6]
```diff
@@ -711,7 +711,7 @@ private String createUidListHeader() {
     }
     
     /**
-     * Takes the name of a message file and returns only the base name.
+     * Takes the name of a message file and returns only the searchBase name.
```
Comment (chibenwa) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java]: oups

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirMessageName.java [hunk 7]
```diff
@@ -343,9 +343,9 @@ public Date getInternalDate() {
     }
     
     /**
-     * Composes the base name consisting of timestamp, unique string and host name
+     * Composes the searchBase name consisting of timestamp, unique string and host name
```
Comment (chibenwa) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirMessageName.java]: oups

File: mpt/core/src/main/java/org/apache/james/mpt/helper/CreateScript.java [hunk 8]
```diff
@@ -356,13 +356,13 @@ public static void rename(ScriptBuilder builder) throws Exception {
         builder.setMailbox(originalMailbox).select();
         builder.setMailbox("moved").select();
         builder.setMailbox("other").select();
-        builder.setMailbox("base").select();
```
Comment (chibenwa) [inline mpt/core/src/main/java/org/apache/james/mpt/helper/CreateScript.java]: ooooouuups. And the followings

**Comment by chibenwa** [comment]
@mbaechler after reading your PR I'm convinced of one thing:

 - We are mixing too much concerns in MailboxQuery. We need to split this up.
 - We 

I believe we should have a simpler interface for **MailboxNameExpression** handling the matching logic. Then we could right a damn simple MailboxQuery.

 - Would make **regex** optional. Yep, JMAP won't do that regex stuff
 - Allow finner grain testing.
 - Makes instanciation easier.

It needs to be extracted in separated class (sorry, I did not understood the logic of yourclass hierarchy). I'm comming up with a demo commit for when you wake up. But you'll see, extracting the regex crazyness makes everything simpler!

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxQuery.java [hunk 9]
```diff
@@ -61,99 +42,186 @@ public static Builder builder(MailboxSession session) {
...
+        }
+
+        @Override
+        public String asString() {
+            return expression;
+        }
+    }
+
+    public enum BuiltinPathExpressionCriteria implements PathExpressionCriteria {
+        MatchAll {
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxQuery.java]: Love that :+1:

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/1011 ...

**Comment by chibenwa** [comment]
Corresponding PR merged.

**Pull request closed without merging**